Buy/Graph.Task.swift

```diff
@@ -26,42 +26,271 @@
+            self.task?.cancel()
         }
         
         // ----------------------------------
-        //  MARK: - Control -
+        //  MARK: - Cache -
         //
+        private func cache(_ data: Data, for hash: Hash) {
+            let cacheItem = CacheItem(hash: hash, data: data)
+            self.cache.set(cacheItem)
+        }
         
-        /// Starts the underlying task
-        public func resume() {
-            self.task.resume()
+        private func cachedModelFor(_ hash: Hash, expireIn: Int? = nil, completion: @escaping (R?) -> Void) {
+            guard let item = self.cache.item(for: hash) else {
+                completion(nil)
+                return
+            }
+            
+            /* ---------------------------------
+             ** If the expiry is provided, we'll
+             ** need to validated that cached
+             ** date isn't older than the allowed
+             ** interval, otherwise return nil.
+             */
+            if let expireIn = expireIn {
+                print("Cache expiry interval set to: \(expireIn)")
+                
+                let now       = Int(Date().timeIntervalSince1970)
+                let timestamp = Int(item.timestamp)
+                
+                guard timestamp + expireIn > now else {
+                    print("Cached item expiry exceeded by: Now: \(timestamp + expireIn - now)")
+                    
+                    /* ----------------------------------
+                     ** Purge any expired items from disk
+                     ** to avoid the overhead of loading
+                     ** them just to find out they have
+                     ** already expired.
+                     */
+                    self.cache.remove(for: hash)
+                    
+                    completion(nil)
+                    return
+                }
+                
+                print("Cached item is still valid. Time remaining Now: \(timestamp + expireIn - now)")
+            } else {
+                print("No cache expiry set.")
+            }
+            
+            let response = HTTPURLResponse(
+                url:          request.url!,
+                statusCode:   200,
+                httpVersion:  "HTTP/1.1",
+                headerFields: nil
+            )
+            let (model, _) = self.processResponse(item.data, response, nil)
+            
+            completion(model)
+        }
+        
+        // ----------------------------------
+        //  MARK: - Session -
+        //
+        private func graphTaskWith(_ request: URLRequest, retryHandler: RetryHandler<R>? = nil, completion: @escaping (R?, Data?, QueryError?) -> Void) -> URLSessionDataTask {
+            return self.session.dataTask(with: request) { data, response, error in
+                
+                let (model, error) = self.processResponse(data, response, error)
+                
+                DispatchQueue.main.async {
```

Comment:
This needs to be executed on the main thread, not the session callback thread. The returned `Task` is not thread-safe (nor is the `Client` for that matter). Anything we return from `Client` is assumed to be on the main thread. Any mutations to `Task`, then, should also be on the main thread.